### DISCUSSION TITLE
Guided Exercise - Circuit Breaker Design Pattern - Spring Microservices - E-Commerce Order Processing System

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/spring-framework#177] [Tasks]
The exercise will be scenario-based, leveraging core Spring Boot concepts, JPA with an H2 database, Rest Template, and Resilience4j for the Circuit Breaker implementation. Spring Actuator will also be used for monitoring. This guide provides a structured, yet complex exercise designed to take a learner a day or more to complete.

#### Overview
Develop an E-Commerce Order Processing System with two primary microservices: `OrderService` and `InventoryService`. The `OrderService` will communicate with the `InventoryService` to check the availability of items. The Circuit Breaker pattern will be implemented to handle failures in the `InventoryService`.

### Code Repository Structure
```
ECommerceSystem/
├── OrderService/
│   ├── src/
│   ├── pom.xml
│   └── application.properties
└── InventoryService/
    ├── src/
    ├── pom.xml
    └── application.properties
```

#### OrderService
- **Responsibilities**: Handle customer orders, communicate with `InventoryService`.
- **Endpoints**:
  - `POST /orders` - Create a new order.
  - `GET /orders/{id}` - Retrieve an order by ID.

#### InventoryService
- **Responsibilities**: Manage inventory data.
- **Endpoints**:
  - `GET /inventory/{itemId}` - Check availability of an item.

### Exercise Objectives

1. **Setup Microservices**: Create `OrderService` and `InventoryService` using Spring Boot.
2. **Implement Basic Operations**:
   - In `OrderService`, implement order creation and retrieval.
   - In `InventoryService`, implement inventory check.
3. **Integrate OrderService with InventoryService**:
   - Use RestTemplate in `OrderService` to call `InventoryService`.
4. **Implement Circuit Breaker**:
   - Integrate Resilience4j with `OrderService`.
   - Configure Circuit Breaker around calls to `InventoryService`.
5. **Simulate Failure**:
   - Introduce a controlled failure in `InventoryService` to trigger the Circuit Breaker.
6. **Implement Monitoring**:
   - Use Spring Actuator to monitor the status of Circuit Breaker.

#### Step 1: Service Setup
- Create Spring Boot projects for `OrderService` and `InventoryService`.
- Configure JPA and H2 database in `application.properties`.

#### Step 2: Business Logic
- Implement the necessary entities, repositories, and controllers for both services.

#### Step 3: Service Communication
- In `OrderService`, use `RestTemplate` to call `InventoryService`.

#### Step 4: Circuit Breaker Integration
- Add Resilience4j dependency in `OrderService`.
- Configure Circuit Breaker around the RestTemplate call to `InventoryService`.
  ```java
  @CircuitBreaker(name = "inventoryService", fallbackMethod = "fallbackMethod")
  public InventoryStatus checkInventory(Long itemId) {
      // RestTemplate call to InventoryService
  }

  public InventoryStatus fallbackMethod(Long itemId, Throwable t) {
      // Fallback logic
  }
  ```

#### Step 5: Simulate and Handle Failures
- Introduce a delay or error in `InventoryService` to simulate failure.
- Ensure that the Circuit Breaker in `OrderService` correctly opens and calls the fallback method.

#### Step 6: Monitoring with Spring Actuator
- Enable Spring Actuator in both services.
- Access Circuit Breaker’s health and metrics endpoints.

### Configurations and Properties

- **OrderService `application.properties`**:
  ```properties
  server.port=8080
  spring.datasource.url=jdbc:h2:mem:ordersdb
  spring.datasource.driverClassName=org.h2.Driver
  spring.datasource.username=sa
  spring.datasource.password=
  spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
  ```
- **InventoryService `application.properties`**:
  ```properties
  server.port=8081
  spring.datasource.url=jdbc:h2:mem:inventorydb
  spring.datasource.driverClassName=org.h2.Driver
  spring.datasource.username=sa
  spring.datasource.password=
  spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
  ```

### Additional Guidelines

- **Version Control**: Use Git for version control. Commit changes frequently with meaningful messages.
- **Testing**: Write unit and integration tests for both services.
- **Error Handling**: Implement global error handling for clean error responses.
- **Documentation**: Document your API endpoints using Swagger or Spring Rest Docs.

### Conclusion

This exercise aims to provide a practical understanding of implementing the Circuit Breaker pattern in a microservices environment using Resilience4j. It covers service integration, fault tolerance, and monitoring, offering a holistic learning experience. By completing this exercise, learners will gain valuable skills in building resilient microservices with Spring Boot.

**Comment by akash-coded** [comment]
Expanding the exercise to include more advanced functionalities of the Circuit Breaker pattern and Resilience4j will provide a deeper understanding of building resilient microservices. We'll add complexities and provide detailed hints to explore the nuanced capabilities of Resilience4j.

### Expanded Exercise: E-Commerce Order Processing System

#### Advanced Objectives

1. **Implement Advanced Circuit Breaker Configurations**:
   - Customize Circuit Breaker thresholds, open and close states, and time windows.
2. **Rate Limiter**:
   - Implement a Rate Limiter for the `InventoryService` calls.
3. **Bulkhead Pattern**:
   - Apply the Bulkhead pattern to isolate parts of `OrderService` from cascading failures.
4. **Retry Mechanism**:
   - Configure a retry mechanism with Resilience4j for transient errors.
5. **Fallback Strategies**:
   - Implement more sophisticated fallback methods for different failure scenarios.
6. **Monitoring and Recovery**:
   - Implement detailed monitoring for Circuit Breaker events and states.
   - Add functionality to recover from different states of the Circuit Breaker.

### Detailed Implementation Steps

#### Step 1: Advanced Circuit Breaker Configurations
- **Hint**: Configure parameters like `failureRateThreshold`, `waitDurationInOpenState`, and `ringBufferSizeInHalfOpenState`.
  ```java
  @CircuitBreaker(name = "inventoryServiceCircuitBreaker", 
                  fallbackMethod = "fallbackMethod")
  public InventoryStatus checkInventory(Long itemId) {
      // RestTemplate call to InventoryService
  }
  ```

#### Step 2: Implement Rate Limiter
- **Hint**: Limit the number of calls to `InventoryService` within a certain time period.
  ```java
  @RateLimiter(name = "inventoryServiceRateLimiter")
  public InventoryStatus checkInventory(Long itemId) {
      // RestTemplate call
  }
  ```

#### Step 3: Apply Bulkhead Pattern
- **Hint**: Use `ThreadPoolBulkhead` or `SemaphoreBulkhead` to limit concurrent calls to certain parts of `OrderService`.
  ```java
  @Bulkhead(name = "inventoryServiceBulkhead")
  public InventoryStatus checkInventory(Long itemId) {
      // RestTemplate call
  }
  ```

#### Step 4: Configure Retry Mechanism
- **Hint**: Use Resilience4j’s `Retry` functionality to retry a call a certain number of times with a specific delay.
  ```java
  @Retry(name = "inventoryServiceRetry")
  public InventoryStatus checkInventory(Long itemId) {
      // RestTemplate call
  }
  ```

#### Step 5: Sophisticated Fallback Methods
- **Hint**: Create different fallback methods for different exceptions or failure scenarios.
  ```java
  public InventoryStatus fallbackMethod(Long itemId, Throwable t) {
      if (t instanceof SpecificException) {
          // Specific fallback logic
      }
      // General fallback logic
  }
  ```

#### Step 6: Monitoring and Recovery
- **Hint**: Use Spring Actuator to monitor the Circuit Breaker status and events. Consider ways to recover or reset the Circuit Breaker state based on the system’s health or after a specific time.

### Additional Configurations and Properties

- **Resilience4j Configuration**:
  - In `application.yml` or `application.properties`, configure the parameters for Circuit Breaker, Rate Limiter, Bulkhead, and Retry.
  - Example:
    ```yaml
    resilience4j.circuitbreaker:
      configs:
        default:
          failureRateThreshold: 50
          waitDurationInOpenState: 10000
          ringBufferSizeInHalfOpenState: 5
          ringBufferSizeInClosedState: 10
    ```

### Additional Guidelines

- **Complex Error Scenarios**: Simulate various types of failures in `InventoryService` such as timeouts, server errors, and network issues.
- **Testing Resilience**: Test how `OrderService` behaves under different failure scenarios. Ensure that it gracefully handles failures, adheres to rate limits, and isolates problems using the Bulkhead pattern.
- **Documentation and Reporting**: Document how different Resilience4j configurations impact the behavior of your services. Use monitoring data to create reports on the resilience of your system.

### Conclusion

This expanded exercise provides a comprehensive learning experience in implementing and managing the Circuit Breaker pattern with Resilience4j in a microservices architecture. By completing this exercise, learners will gain in-depth knowledge about building resilient systems capable of handling various failure modes gracefully.

**Comment by akash-coded** [comment]
To further expand and enrich the exercise with a focus on more complex and creative implementations of the Circuit Breaker pattern using Resilience4j, we'll delve into advanced concepts and scenarios. This expansion will explore intricate uses of the Circuit Breaker and its interaction with other resilience mechanisms, adding depth and complexity to the exercise.

### Advanced E-Commerce Order Processing System

#### Expanded Advanced Objectives

1. **Dynamic Circuit Breaker Configuration**:
   - Implement functionality to adjust Circuit Breaker parameters at runtime based on system behavior or external triggers.

2. **Circuit Breaker State Persistence**:
   - Persist the state of the Circuit Breaker to a database or a distributed cache to maintain state across service restarts.

3. **Complex Fallback Mechanisms**:
   - Develop advanced fallback strategies, such as invoking alternative microservices or providing default responses based on specific failure types.

4. **Circuit Breaker Events and Analytics**:
   - Capture and analyze Circuit Breaker events for insights into system performance and failure patterns.

5. **Integration with Message Queues**:
   - In case of service unavailability, use message queues to buffer requests and process them when the service recovers.

6. **A/B Testing with Circuit Breakers**:
   - Use Circuit Breakers to perform A/B testing by selectively routing traffic to different service versions based on certain criteria.

### Detailed Implementation Steps

#### Step 1: Dynamic Circuit Breaker Configuration
- **Hint**: Utilize external configuration tools or custom endpoints to modify Circuit Breaker settings dynamically.
- **Implementation**: Create an admin endpoint in `OrderService` to adjust Circuit Breaker parameters. Use the Resilience4j API to apply these changes at runtime.

#### Step 2: Circuit Breaker State Persistence
- **Hint**: Use a database or distributed cache like Redis to store Circuit Breaker states.
- **Implementation**: Integrate a persistence mechanism to store and retrieve the Circuit Breaker’s state.

#### Step 3: Complex Fallback Mechanisms
- **Hint**: Design fallback methods to handle different types of failures intelligently.
- **Implementation**: Implement multiple fallback methods in `OrderService` that are selected based on the exception type or failure scenario.

#### Step 4: Circuit Breaker Events and Analytics
- **Hint**: Use event listeners to capture Circuit Breaker events and store them for analysis.
- **Implementation**: Implement an event logging mechanism within Resilience4j’s Circuit Breaker to track its state changes and decisions.

#### Step 5: Integration with Message Queues
- **Hint**: Use a message queue like RabbitMQ or Kafka to buffer requests during downtime.
- **Implementation**: Configure `OrderService` to send requests to a queue when `InventoryService` is down, and process them later.

#### Step 6: A/B Testing with Circuit Breakers
- **Hint**: Route requests to different service instances based on Circuit Breaker status or other criteria.
- **Implementation**: Set up two versions of `InventoryService` and use the Circuit Breaker’s status or other metrics to route requests for A/B testing.

### Additional Configurations and Properties

- **Dynamic Configuration**:
  - Implement endpoints in `OrderService` to receive and apply new Circuit Breaker configurations.
- **Event Logging and Analytics**:
  - Set up a logging framework to capture and store Circuit Breaker events for further analysis.

### Additional Guidelines

- **Simulating Advanced Scenarios**: Create scenarios where the system's behavior changes dynamically, requiring adjustments to the Circuit Breaker settings.
- **Monitoring and Alerting**: Integrate monitoring tools to track the health of microservices and set up alerts based on Circuit Breaker events.
- **Documentation**: Update documentation to cover the advanced functionalities and scenarios implemented.

### Conclusion

This expanded and enriched exercise offers a comprehensive and deep dive into the Circuit Breaker pattern using Resilience4j, pushing the boundaries of conventional implementations. By exploring dynamic configurations, state persistence, complex fallback mechanisms, analytics, integration with message queues, and creative uses like A/B testing, learners will gain a profound understanding of building highly resilient and adaptable microservices. The exercise not only imparts technical skills but also fosters a mindset for innovative problem-solving in distributed systems.

**Comment by samik2023** [comment]
Hi Akash,

Here is the resilience4j changes.

https://github.com/samik2023/order-service
https://github.com/samik2023/book-service

Regards
Samik

**Comment by arulsudarvizhi1** [comment]
Hi Akash ,
@akash-coded 
Please find attached the code for circuit breaker with resilience implementation. Tested the changes by not starting the inventory service where default the fallback method was called.
inventory-service
order-service: Contains book implementation: http://localhost:8083/order  [Api whose service have the circuit breaker implementation]

[circuitbreaker.zip](https://github.com/akash-coded/spring-framework/files/13429226/circuitbreaker.zip)

Name: Arul
Batch: Nov 7th to Dec 6th  7.30 am

**Comment by sowmyagardas** [comment]
Hi Akash,
I have implemented the circuit breaker design pattern and attached the code base. Please check.

Thanks,
Sowmya.
[Circuitbreaker_Pattern_Sowmya.zip](https://github.com/akash-coded/spring-framework/files/13433540/Circuitbreaker_Pattern_Sowmya.zip)

**Comment by HarshDPandya** [comment]
Hi @akash-coded , Sharing my code in zip file for **Guided Exercise - Circuit Breaker Design Pattern - Spring Microservices - E-Commerce Order Processing System**
Name: Harsh Pandya
TCS Emp Id: 2066481
[microservice-ecommerce-circuit-breaker-HarshPandya-2066481.zip](https://github.com/akash-coded/spring-framework/files/13449425/microservice-ecommerce-circuit-breaker-HarshPandya-2066481.zip)

**Comment by vishalakshi2** [comment]
Hi Akash,

Implemented the circuit breaker pattern in e-commerce app. Please find the code for the same:
[ecommerce-app.zip](https://github.com/akash-coded/spring-framework/files/13454214/ecommerce-app.zip)
Tcs Emp Id: 922774

**Comment by himarou** [comment]
Hi Akash,

I have implemented circuit breaker in e-commerce Microservices. Please find the code attached.
[ecommerce-microservice-circuit-breaker.zip](https://github.com/akash-coded/spring-framework/files/13464685/ecommerce-microservice-circuit-breaker.zip)
TCS EmployeeID - 334368

Thanks,
Himansu Rout

**Comment by ssagarwala** [comment]
Hello Akash,  @akash-coded 
good morning, 

Please check my work in this repo https://github.com/ssagarwala/CircuitBreakerExercise. 

Thanks and Regards, 
-Fatima.

**Comment by AnubhavG1999** [comment]
Hi Akash,

Here's my circuit breaker implementation, apologies for the delay. Thanksgiving weekend got the better of me!

Name: Anubhav Ghosh
Org: TCS
Emp ID: 2286865

[CircuitBreaker.zip](https://github.com/akash-coded/spring-framework/files/13476841/CircuitBreaker.zip)

**Comment by pradnyakinge** [comment]
hi @akash-coded , I have implemented circuit breaker in prior assignment code . Below is the link of implementing circuit breaker ad postman collection : https://github.com/pradnyakinge/SpringBoot-Microservices/tree/main/CircuitBreakerInOrderService

The above project contain all api gateway, error handling, circuit breaker for feign calls. I hope that will work for you otherwise let me know , I will create new microservices suggested in this task and apply circuit breaker in it.

name :Pradnya Kinge
tcs id: 1052359

**Comment by obulix** [comment]
Hi @akash-coded ,

Here is the circuit breaker implementation.
https://github.com/obulix/MicSrv-CircuitBreaker
Name : Dinakaran

**Comment by jyothi241868** [comment]
Hi Akash,

Please find the attached for Circuit Breaker implementation.

[CircuitBreaker.zip](https://github.com/akash-coded/spring-framework/files/13601502/CircuitBreaker.zip)


Thanks,
Jyothi K
